src/server/dflycmd.h

```diff
@@ -29,29 +31,91 @@ namespace journal {
 class Journal;
 }  // namespace journal
 
+// DflyCmd is responsible for managing replication. A master instance can be connected
+// to many replica instances, what is more, each of them can open multiple connections.
+// This is why its important to understand replica lifecycle management before making
+// any crucial changes.
+//
+// A SyncInfo instance is responsible for managing a replica's state and is accessible by its
+// sync_id. Each per-thread connection is called a Flow and is represented by the FlowInfo
+// instace, accessible by its index.
+//
+// An important aspect is synchronization and efficient locking. Three levels of locking are used:
+//  1. Global locking.
+//    Member  mutex `mu_` is used for synchronizing operations connected with internal data
+//    structures.
+//  2. Per-replica locking
+//    SyncInfo contains a separate mutex that is used for replica-only routines. It is held during
+//    state transitions (start full sync, start stable state sync) and mebmer access.
+//  3. Connection per-replica locking.
+//    Once we aborted the replication process, we need to keep Dragonfly's connection objects alive
+//    during the cleanup phase by blocking in the OnClose call (because internal structures like
+//    RdbSaver still keep references to them). This is done by separate connection_hold locks inside
+//    SyncInfo
+//
```

Comment:
In theory, we can use the per-replica lock for exactly the same purpose. It'll make the code easier indeed.
But it will also make the replicas life a lot harder by long holding command that should have failed fastly.

However, we can solve the issue with the second approach by re-using the mutex and making the state an seq-cst atomic, so that _before_ trying to lock the mutex, we can check wether it even makes sense by reading the atomic (because a CANCELLED state can't be un-cancelled). We should though check a second time after we acquired the lock to see if it didn't change since then